truss/remote/baseten/remote.py

```diff
@@ -175,44 +188,173 @@ def push(  # type: ignore
+        promote: bool = False,
+        preserve_previous_prod_deployment: bool = False,
+        disable_truss_download: bool = False,
+        deployment_name: Optional[str] = None,
+        origin: Optional[custom_types.ModelOrigin] = None,
+        environment: Optional[str] = None,
+        chain_environment: Optional[str] = None,
+        chainlet_name: Optional[str] = None,
+        chain_name: Optional[str] = None,
+    ) -> BasetenService:
+        push_data = self._prepare_push(
+            truss_handle=truss_handle,
+            model_name=model_name,
+            publish=publish,
+            trusted=trusted,
+            promote=promote,
+            preserve_previous_prod_deployment=preserve_previous_prod_deployment,
+            disable_truss_download=disable_truss_download,
             deployment_name=deployment_name,
             origin=origin,
             environment=environment,
+        )
+
+        model_id, model_version_id = create_truss_service(
+            api=self._api,
+            model_name=push_data.model_name,
+            s3_key=push_data.s3_key,
+            config=push_data.encoded_config_str,
+            is_draft=push_data.is_draft,
+            model_id=push_data.model_id,
+            is_trusted=push_data.is_trusted,
+            preserve_previous_prod_deployment=push_data.preserve_previous_prod_deployment,
+            allow_truss_download=push_data.allow_truss_download,
+            deployment_name=push_data.version_name,
+            origin=push_data.origin,
+            environment=push_data.environment,
             chain_environment=chain_environment,
             chainlet_name=chainlet_name,
             chain_name=chain_name,
-            allow_truss_download=not disable_truss_download,
         )
 
         return BasetenService(
             model_id=model_id,
             model_version_id=model_version_id,
-            is_draft=not publish,
+            is_draft=push_data.is_draft,
             api_key=self._auth_service.authenticate().value,
             service_url=f"{self._remote_url}/model_versions/{model_version_id}",
             truss_handle=truss_handle,
             api=self._api,
         )
 
+    def push_chain_atomic(
+        self,
+        chain_name: str,
+        chainlet_artifacts: List[custom_types.ChainletArtifact],
+        publish: bool = False,
+        environment: Optional[str] = None,
+    ):
+        # If we are promoting a model to an environment after deploy, it must be published.
+        # Draft models cannot be promoted.
+        if environment and not publish:
+            logging.info(
+                f"Automatically publishing Chain '{chain_name}' based on environment setting."
+            )
+            publish = True
+
+        chain_id = get_chain_id_by_name(self._api, chain_name)
+        chainlet_data: List[custom_types.ChainletDataAtomic] = []
+        entrypoint_truss_handle: Optional[TrussHandle] = None
+
+        for artifact in chainlet_artifacts:
```

Comment:
Actually I don't think parallelizing uploads will be much faster. Each upload is happening with multiple threads anyway per [this](https://github.com/basetenlabs/truss/blob/7d9e95bd4e9e3530073e8a09428b11a34f207e43/truss/remote/baseten/utils/transfer.py#L33-L34).